### DISCUSSION TITLE
How to get self Hosted Docker to render GitHub badges

### DISCUSSION THREAD

**Discussion by Elaniobro** [badges/shields#7106] [Q&A] [answered]
I have a server that I am currently using for github actions with an on prem github enterprise server. I wanted to add to the github runner server the shields.io self hosted applicaiton. I have done this via the docker route.

However, when I try to generate badges, I consistently get invalid badges, see screen shot below:
<img width="825" alt="Screen Shot 2021-10-07 at 17 12 48" src="https://user-images.githubusercontent.com/710847/136463305-d3668157-9786-4c7b-aa80-6aec7f319424.png">
:

I am currently passing the environment variables to the `docker run` command like so:

`docker run --rm -d -e GITHUB_URL='https://onprem.github.com' -e GH_TOKEN='MYTOKEN' -p 127.0.0.1:8080:80 --name shields shieldsio/shields:next`

_ip, token and url were changed for privacy_
Do I need to add other variables and or set some more options for this to work?

edit: When looking at my Enterprise personal access token, it says it was never used....

**Comment by calebcartwright** [comment]
Sounds like another case of #6910. You'd probably want to do `https://onprem.github.com/api/v3` as the value for GITHUB_URL for all of the v3-based badges, but worth noting that our issue badge takes advantage of the v4/graphql endpoint instead (which would require a different value)

**Reply by Elaniobro** [reply]
@calebcartwright thank you for your reply. Updating the endpoint does in fact remove the invalid tag. However, I am now seeing `issues: not found` which is not correct as my repo has quite a few.

Do I need to update to use the `v4/graphql` endpoint? 
And by different values:

> (which would require a different value)

are you referring to the URL or the inputs for which I add to generate the badge, those being: `user` and `repo` in the example blow?
<img width="812" alt="Screen Shot 2021-10-11 at 17 49 03" src="https://user-images.githubusercontent.com/710847/136859761-1c359265-34e8-4c32-8a95-3208db24c7cd.png">

**Reply by Elaniobro** [reply]
FWIW I have tried swapping out the URL:

`https://myserver.com/api/v3` with `https://myserver.com/api/graphql` neither works for the issue(s) badge(s). 

I did notice that _some_ badges work when the `/api/v3`URL is passed into the docker build.

**Reply by calebcartwright** [reply, marked as answer]
We provide a _lot_ of GitHub related badges (I believe the most of any service), which means we interface with a decent chunk of GitHub's API surface. Historically, we'd exclusively relied on GitHub's Rest/v3 APIs, but in more recent times have added new badges and even converted a couple of the older ones to use the GraphQL/v4 endpoint (sometimes because it's required, and other times because it was beneficial for us to do so).

That works great for GitHub's cloud/SaaS offerings, including Enterprise Cloud. However, as in noted #6910, it's a challenge for users that are self-hosting a Badge server and trying to generate badges that are pulling from a GitHub Enterprise Server (ES) instance because ES does not follow the same uri patterns.

This means that folks in that dual self-hosting scenario currently have to pick one subset of our GitHub badges, or the other; specifically, either the subset of our GitHub badges which internally utilize GitHub's rest/v3 endpoints or the other subset that utilizes GraphQL/v4 (or I suppose alternatively to the alternative, one could host two different badge servers, respectively configured with the two different GitHub endpoint values).

If you want to work with the GraphQL-dependent based badges with ES, which includes the Issues badge you've referenced, then you'd set your badge server config to `https://myserver.com/api`. If instead you want to other set of REST-dependent badges with ES, then you'd have to set your config to `https://myserver.com/api/v3`.

There's not a streamlined way of determining which of our GitHub badges are dependent on which API surface, so to make that determination you'd have to look at the underlying code to find the base class for the corresponding service class (the service class and filenames should be named close enough to the badge to identify them, but it's not necessarily an exact science).

e.g.

The GitHub Contributors badge uses the v3/rest endpoint (evident by the base class `GithubAuthV3Service`):
https://github.com/badges/shields/blob/4c8a211febc792a67319fcbc70c73be4681c4a26/services/github/github-contributors.service.js#L10

whereas the GitHub Deployments badge uses the v4/graphql endpoint (evident by the base class `GithubAuthV4Service`):
https://github.com/badges/shields/blob/4c8a211febc792a67319fcbc70c73be4681c4a26/services/github/github-deployments.service.js#L45
